### PR TITLE
Feature/web-application-factory

### PR DESCRIPTION
**Pull request by kashifsoofi**
#21 Updated intergration tests to use WebAppliationFactory

- Updated IntegrationFixture to create InMemory web server using WebApplicationFactory
- Updated integration tests
- Added HttpMethodMiddleware as InMemory TestHost was not filtering body in HEAD requests
- Updated WebSocket test to use WebSocketClient from WebApplicationFactory.TestHost

### CODE REVIEW

File: FakeServer/Startup.cs [hunk 1]
```diff
@@ -130,6 +130,8 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, IApplica
 
             app.UseMiddleware<HttpOptionsMiddleware>();
 
+            app.UseHeadMethodMiddleware();
```
Comment (ttu) [inline FakeServer/Startup.cs]: Normally Kestrel takes care of cleaning up the body in HEAD cases, so this needs to be only used when running the tests

```cs
if (env.EnvironmentName == "IntegrationTest")
            {
                app.UseHeadMethodMiddleware();
            }
```

Better way would be to add this when WebApplicationFactory is created, but no idea how to do that. Maybe it is not even possible 😄

File: FakeServer.Test/FakeServerSpecs.cs [hunk 2]
```diff
@@ -221,586 +185,487 @@ public async Task GetUsers_PageDoesNotExist_ReturnsEmptyList()
...
-                Assert.Equal(HttpStatusCode.NotFound, result.StatusCode);
-            }
+            result = await _fixture.Client.GetAsync($"api/{collection}");
+            Assert.Equal(HttpStatusCode.NotFound, result.StatusCode);
         }
 
         [Fact]
         public async Task WebSockets_CalledTwice()
         {
             var are = new AutoResetEvent(false);
```
Comment (ttu) [inline FakeServer.Test/FakeServerSpecs.cs]: AutoResetEvent is not needed anymore with this implementation. 

Just use
```cs
await WebSocketReceiveHandler();
```
and you can take remove AutoResetEvent.

File: FakeServer/Common/HeadMethodMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            {
+                context.Request.Method = HttpMethods.Head;
+            }
+        }
+    }
+
+    // ref: https://www.tutorialsteacher.com/core/how-to-add-custom-middleware-aspnet-core
+    public static class HeadMethodMiddlewareExtensions
+    {
+        public static IApplicationBuilder UseHeadMethodMiddleware(this IApplicationBuilder builder)
```
[the remark this answers is not in the document]
Comment (kashifsoofi) [reply FakeServer/Common/HeadMethodMiddleware.cs]: Done, removed.

File: FakeServer/Common/HeadMethodMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    public class HeadMethodMiddleware
+    {
+        private readonly RequestDelegate _next;
+
+        public HeadMethodMiddleware(RequestDelegate next)
+        {
+            _next = next;
+        }
+
+        public async Task Invoke(HttpContext context)
```
[the remark this answers is not in the document]
Comment (kashifsoofi) [reply FakeServer/Common/HeadMethodMiddleware.cs]: Thanks for the suggestion, sounds good, done.

File: FakeServer/Startup.cs (hunk 1 above)
Comment (kashifsoofi) [reply FakeServer/Startup.cs]: Done.

File: FakeServer.Test/FakeServerSpecs.cs (hunk 2 above)
Comment (kashifsoofi) [reply FakeServer.Test/FakeServerSpecs.cs]: Done.

File: FakeServer.Test/IntegrationTestCollection.cs [hunk 5]
```diff
@@ -1,92 +1,102 @@
...
+                mainConfiguration.Add("Authentication:AuthenticationType", authenticationType);
+                mainConfiguration.Add("Authentication:Users:0:Username", "admin");
+                mainConfiguration.Add("Authentication:Users:0:Password", "root");
+            }
 
-            var success = Task.Run(() => WaitForServer()).GetAwaiter().GetResult();
+            _factory = new WebApplicationFactory<Startup>()
+                .WithWebHostBuilder(builder =>
+                {
+                    builder.UseEnvironment("Development")
```
[the remark this answers is not in the document]
Comment (kashifsoofi) [reply FakeServer.Test/IntegrationTestCollection.cs]: Done.

File: FakeServer/Startup.cs (hunk 1 above)
Comment (kashifsoofi) [reply FakeServer/Startup.cs]: Maybe raise a separate issue for improvement.

**Comment by kashifsoofi** [comment]
Changes done as per suggestions.

File: FakeServer/Startup.cs (hunk 1 above)
Comment (ttu) [reply FakeServer/Startup.cs]: Yes, this is a case for an own issue 👍

**Comment by ttu** [comment]
Thanks 👍

**Pull request merged**